### PR TITLE
Add images with flash attention 2

### PR DESCRIPTION
**Pull request by dakinggg**
CI passes when run manually with flash attention 2
```
=============================================================================== 471 passed, 2 skipped, 239 deselected, 73 xfailed, 144 warnings in 309.04s (0:05:09) ================================================================================
pip show flash-attn
Name: flash-attn
Version: 2.3.2
Summary: Flash Attention: Fast and Memory-Efficient Exact Attention
Home-page: https://github.com/Dao-AILab/flash-attention
Author: Tri Dao
Author-email: trid@cs.stanford.edu
License: 
Location: /usr/lib/python3/dist-packages
Requires: einops, ninja, packaging, torch
Required-by: 

```

Will do more testing after this PR goes in so that I can add proper CI and use properly built images instead of just test ones.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -91,6 +91,12 @@
     # PyPI does not support direct dependencies, so we remove this line before uploading from PyPI
     'xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v1.0.9#subdirectory=csrc/xentropy',
 ]
+extra_deps['gpu-flash2'] = [
+    'flash-attn==2.3.2',
+    'mosaicml-turbo==0.0.4',
+    # PyPI does not support direct dependencies, so we remove this line before uploading from PyPI
+    'xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v2.3.1#subdirectory=csrc/xentropy',
```
Comment (vchiley) [inline setup.py:98]: why is `'flash-attn==2.3.2'` but `'xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v2.3.1#subdirectory=csrc/xentropy'`

ie 2.3.2 vs 2.3.1

Comment (dakinggg) [reply setup.py:98]: cause I made a typo

**Review by vchiley — APPROVED** [review]

File: .github/workflows/docker.yaml [hunk 2]
```diff
@@ -52,13 +64,32 @@ jobs:
...
+        echo "IMAGE_CACHE=${IMAGE_CACHE}" >> ${GITHUB_ENV}
+
     - name: Build and Push the Docker Image
       uses: docker/build-push-action@v3
       with:
         context: .
-        tags: mosaicml/llm-foundry:${{ matrix.name }}-latest,
-          mosaicml/llm-foundry:${{ matrix.name }}-${{ env.IMAGE_TAG }}
+        tags: ${{ env.IMAGE_TAG }}
         push: true
```
Comment (mvpatel2000) [inline .github/workflows/docker.yaml:90]: Are you getting rid of the latest tag?

**Review by mvpatel2000 — APPROVED** [review]
It's hard to reason about the docker changes. I am assuming you tested this

File: .github/workflows/docker.yaml (hunk 2 above)
Comment (dakinggg) [reply .github/workflows/docker.yaml:90]: No, the latest is baked in to the image tag env var set above. This is how we set it in composer too.

**Comment by dakinggg** [comment]
@mvpatel2000 I tested in that working images get pushed to the staging repo from this pr. I suppose I haven't tested that I didn't break the push to main workflow that releases the production images.

**Pull request merged**